x-pack/plugins/security_solution/server/endpoint/services/actions.ts

```diff
@@ -211,37 +263,51 @@ export const getPendingActionCounts = async (
           query: {
             bool: {
               filter: [
-                { terms: { action_id: actionIDs } }, // get results for these actions
-                { terms: { agent_id: agentIDs } }, // ignoring responses from agents we're not looking for
+                { terms: { action_id: actionIds } }, // get results for these actions
+                { terms: { agent_id: agentIds } }, // ONLY responses for the agents we are interested in (ignore others)
               ],
             },
           },
         },
       },
       { ignore: [404] }
     )
-    .then((result) => result.body?.hits?.hits?.map((a) => a._source!) || []);
-
-  // respond with action-count per agent
-  const pending: EndpointPendingActions[] = agentIDs.map((aid) => {
-    const responseIDsFromAgent = responses
-      .filter((r) => r.agent_id === aid)
-      .map((r) => r.action_id);
-    return {
-      agent_id: aid,
-      pending_actions: recentActions
-        .filter((a) => a.agents.includes(aid) && !responseIDsFromAgent.includes(a.action_id))
-        .map((a) => a.data.command)
-        .reduce((acc, cur) => {
-          if (cur in acc) {
-            acc[cur] += 1;
-          } else {
-            acc[cur] = 1;
-          }
-          return acc;
-        }, {} as EndpointPendingActions['pending_actions']),
-    };
-  });
+    .then((result) => result.body?.hits?.hits?.map((a) => a._source!) || [])
+    .catch(catchAndWrapError);
 
-  return pending;
+  if (actionResponses.length === 0) {
+    return actionResponsesByAgentId;
+  }
+
+  // Get the latest docs from the metadata datastream for the Elastic Agent IDs in the action responses
+  // This will be used determine if we should withhold the action id from the returned list in cases where
+  // the Endpoint might not yet have sent an updated metadata document.
+  const latestEndpointMetadataDocs = await metadataService.findHostMetadataForFleetAgents(
+    esClient,
+    agentIds
+  );
+
+  // Object of Elastic Agent Ids to event created date
+  const endpointLastEventCreated: Record<string, Date> = latestEndpointMetadataDocs.reduce(
+    (acc, endpointMetadata) => {
+      acc[endpointMetadata.elastic.agent.id] = new Date(endpointMetadata.event.created);
+      return acc;
+    },
+    {} as Record<string, Date>
+  );
+
+  for (const actionResponse of actionResponses) {
+    const lastEndpointMetadataEventTimestamp = endpointLastEventCreated[actionResponse.agent_id];
+
+    // FIXME: need to account for the action expiration time or include some arbitrary check here so that we don't get potential pending actions forever.
```

Comment:
only unexpired actions are retrieved in the first call in the 'handler', when getting the action IDs

Reply:
thats right - thanks for the reminder.
Take a look at the current implementation. I added an "elapsed time" to it of 5 minutes just to protect against the possibility of perhaps kibana never receiving a metadata document from the endpoint. So it will try to see if we received a metadata document only for a duration of 5 minutes after